### ISSUE TITLE
Exception loading helper: /packages/velocity-jasmine-unit/lib/loader-helper.js

### ISSUE THREAD

**Issue by gerwinbrunner** [xolvio/jasmine-unit#6]
Hi I just played around with velocity and velocity-jasmine-unit.

I got mocha-web to work, but when I have test-jasmine-unit.js in the test folder I get the following error in the server console:

I20140513-23:50:16.615(2)? mocha-web: First test executed
I20140513-23:52:38.745(2)? Exception loading helper: /Users/ap/Documents/workspaces/VilangoWork/vicoding-v3/packages/velocity-jasmine-unit/lib/loader-helper.js
I20140513-23:52:38.746(2)? [ReferenceError: _ is not defined]

Any ideas what I'm doing wrong here?

Cheers,
Gerwin

**Comment by SamHatoum** [comment]
Hi.

Are you using underscore in your app?

**Comment by SamHatoum** [comment]
or lodash

**Comment by gerwinbrunner** [comment]
No not explicitly.

Here are my **packages** file:

```
standard-app-packages
autopublish
insecure
velocity
mocha-web
coffeescript
velocity-jasmine-unit
moment
bootstrap-3
npm
iron-router
jquery-autosize
tv4
chai
```

and my **smart.json** file:

```
{
  "packages": {
    "velocity": {
      "git": "https://github.com/xolvio/velocity"
    },
    "mocha-web": {
      "git": "https://github.com/mad-eye/meteor-mocha-web",
      "branch": "mtr"
    },
    "velocity-jasmine-unit": {
      "git": "https://github.com/xolvio/velocity-jasmine-unit"
    },
    "moment": {},
    "bootstrap-3": {},
    "tv4": {},
    "jquery-autosize": {},
    "npm": {},
    "iron-router": {},
    "chai": {}
  }
}

```

**Comment by gerwinbrunner** [comment]
and my **packages.json** file:

```
{
  "simple-text-parser": "0.1.2",
  "xregexp": "2.0.0"
}
```

**Comment by alanning** [comment]
`velocity-jasmine-unit` itself uses lodash:
- [package.js](https://github.com/xolvio/velocity-jasmine-unit/blob/master/package.js#L7)
- [main.js](https://github.com/xolvio/velocity-jasmine-unit/blob/master/main.js#L12)
- [lib/coffee-require.js](https://github.com/xolvio/velocity-jasmine-unit/blob/master/lib/coffee-require.js#L10)
- [lib/file-loader.js](https://github.com/xolvio/velocity-jasmine-unit/blob/master/lib/file-loader.js#L9)
- [lib/html-scanner.js](https://github.com/xolvio/velocity-jasmine-unit/blob/master/lib/html-scanner.js#L19)
- [lib/load-order-sort.js](https://github.com/xolvio/velocity-jasmine-unit/blob/master/lib/load-order-sort.js#L7)

`_` is defined in all instances so maybe the lodash package itself isn't getting installed by the time the code in question is run?

If you run `npm install` in the `velocity-jasmine-unit` directory, does that resolve the issue?

**Comment by gerwinbrunner** [comment]
When I run `npm install` in `packages/velocity-jasmine-unit` of my project then I get the following output:

```
npm ERR! install Couldn't read dependencies
npm ERR! package.json ENOENT, open '/Users/ap/.meteorite/packages/velocity-jasmine-unit/xolvio/velocity-jasmine-unit/47173dc3e6bae823b3a20b93382fe68a3f1e4b11/package.json'
npm ERR! package.json This is most likely not a problem with npm itself.
npm ERR! package.json npm can't find a package.json file in your current directory.

npm ERR! System Darwin 13.1.0
npm ERR! command "node" "/usr/local/bin/npm" "install"
npm ERR! cwd /Users/ap/.meteorite/packages/velocity-jasmine-unit/xolvio/velocity-jasmine-unit/47173dc3e6bae823b3a20b93382fe68a3f1e4b11
npm ERR! node -v v0.10.28
npm ERR! npm -v 1.4.10
npm ERR! path /Users/ap/.meteorite/packages/velocity-jasmine-unit/xolvio/velocity-jasmine-unit/47173dc3e6bae823b3a20b93382fe68a3f1e4b11/package.json
npm ERR! code ENOPACKAGEJSON
npm ERR! errno 34
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /Users/ap/.meteorite/packages/velocity-jasmine-unit/xolvio/velocity-jasmine-unit/47173dc3e6bae823b3a20b93382fe68a3f1e4b11/npm-debug.log
npm ERR! not ok code 0
```

**Comment by alanning** [comment]
Hi @gerwinbrunner,

I'm having trouble reproducing the issue.  We just pushed a new version of velocity to atmosphere so please have a try with [v0.1.1](https://atmospherejs.com/package/velocity) and see if that resolves the issue.  Maybe just something out-of-date.

If the issue still arrises, please post a repro to github and I'll debug further.

One other thing to try is running meteor with the `DEBUG=1` environment variable set.  It will output additional debug info for `velocity-jasmine-unit`.  I use a bash script but I think you can do it this way:

```
$ DEBUG=1 meteor
```

**Comment by gerwinbrunner** [comment]
Hi @alanning 

I started to do some debugging and create a sample app to make it as easy as possible for you to get the actual error.
While creating the sample app I figured out that this issue has to do with the [scaffolding 'em'](https://github.com/EventedMind/em) tool I was using.

You can checkout the [sample app](https://github.com/Vilango/meteor-sample-app) and start it.
You'll run immediately into the following error:

```
I20140516-16:09:35.483(2)? Exception loading helper: /Users/ap/Documents/workspaces/github-workspace/meteor-sample-app/packages/velocity-jasmine-unit/lib/loader-helper.js
I20140516-16:09:35.484(2)? [ReferenceError: Router is not defined]
```

So I assume this issue is related to the issue #3.

Is there a way to tell velocity to ignore certain files while the stubbing is not there?

**Comment by gerwinbrunner** [comment]
@samhatoum the answer to your question is yes.

**Comment by alanning** [comment]
Hi @gerwinbrunner,

Thanks for researching this further.

In order to get your app running with unit tests, you'll have to stub/mock any packages that are used in Meteor.startup or the specific functions you are trying to test.  (Once we have auto-mocking working you won't have to do that manually anymore.)

I'm curious if you'll still see the `_` error once the mocking for router is in place.  If you do, please update your repro.  If not, then go ahead and close this issue as it's already covered by #3.

**Comment by gerwinbrunner** [comment]
Hi @alanning,

I added the sub and now the initial error shows up again.

```
I20140519-15:21:14.322(2)? Exception loading helper: /Users/ap/Documents/workspaces/github-workspace/meteor-sample-app/packages/velocity-jasmine-unit/lib/loader-helper.js
I20140519-15:21:14.323(2)? [ReferenceError: _ is not defined]
```

Here is the updated [sample app](https://github.com/Vilango/meteor-sample-app).

I guess the code in generated by 'em' (client/app.js) needs underscore.js and it is not yet loaded.

**Comment by gerwinbrunner** [comment]
Any hints on this?

**Comment by alanning** [comment]
Hi @gerwinbrunner,

In your `client/app.js` file you use `_.extend` and `_.each` so those need to be stubbed before the unit tests can run.

I submitted a PR to your example app repo which fixes the issue.

EDIT:  This won't be necessary once the auto-stubbing of packages is implemented.  Till then, I'll see about adding a specific underscore stub to jasmine-unit since that's a very commonly used package.

**Comment by gerwinbrunner** [comment]
Oh, that makes sense. 
I was thinking on how to get underscore loaded :)

Thanks for applying the patch.

P.S.: The improvements to velocity (stacktracke are awesome!) Thanks!

**Comment by queso** [comment]
I too am getting this error.  Isn't autostubber in the package now?

**Comment by MichalW** [comment]
Hi, I have some problem with loading helper. I have global variables in 'use strict' mode.

file1.js:

``` javascript
'use strict';

(Meteor.isClient ? window : global).ExampleFunction = function (){
    return true;
};
```

file2.js

``` javascript
'use strict';

ExampleFunction();
```

Logs:

```
1404319597191   out Exception loading helper: /home/michal/public_html/Vazco/meteor/test_velocity2/packages/jasmine-unit/lib/loader-helper.js   jasmine-unit
1404319597201   out [ReferenceError: ExampleFunction is not defined]    jasmine-unit
1404319597361   out Exception loading helper: /home/michal/public_html/Vazco/meteor/test_velocity2/packages/jasmine-unit/lib/loader-helper.js [ReferenceError: ExampleFunction is not defined]  jasmine-unit
```

**Comment by alanning** [comment]
Hi @MichalW, I'm not sure why this doesn't work.  It appears its something to do with how jasmine-cli does things as jasmine-unit is actually kicking off jasmine-cli under the hood.  Run your app with full debugging output turned on and try the jasmine-cli command that is executed and you'll see the same error.

For me that's:

``` bash
$ DEBUG=1 JASMINE_DEBUG=1 VELOCITY_DEBUG=1 mrt
```

Not sure how jasmine cli does globals.

The only way I could get it to work was to define ExampleFunction above the `'use strict';` like so:

``` js
ExampleFunction = function () {
  console.log('example function executed');
  return true;
};

'use strict';
```

**Comment by MichalW** [comment]
I do not understand what the problem is, but I found solution:

``` javascript
'use strict';

(typeof global !== 'undefined'? global : window).ExampleFunction = function (){
    return true;
};
```

**Comment by alanning** [comment]
Nice job finding that!  So jasmine-cli handles global after all, it just also provides a window which was messing the other code up.

**Comment by MichalW** [comment]
This is funny, because it works too:

``` javascript
'use strict';

(Meteor.isServer ? global : window).ExampleFunction = function (){
    return true;
};
```

but this not:

``` javascript
'use strict';

(typeof window !== 'undefined'? window : global).ExampleFunction = function (){
    return true;
};
```

**Comment by alanning** [comment]
It looks like node or jasmine is supplying a 'window' object.  In the code that doesn't work, the default is window which is apparently not the same as the root (global) context in nodejs apps. 

The failing case in file2.js can actually be written as global.ExampleFunction, so since window is a different obj, it throws the error.

**Comment by MichalW** [comment]
Second problem, that I have found, is loading order of js files. Jasmine loads them alphabetically and if I write the definition of variable to the file in `lib` directory, it not load earlier.
